### PR TITLE
[ready] expanded tests & bugfixes

### PR DESCRIPTION
**Pull request by sbma44**
refs #18 and #19 

- adds offline tests for gzips, tarballs and purely-theoretical multi-mmdb tarballs (basically just for the sake of code coverage)
- reinserts `use strict` to allow node 4 compatibility
- fixes a bug in tarball handling that prevented callbacks from firing properly
- adjusts behavior of tarball function to respect the `local` value in `config.js`, bringing it more in line with the project's earlier code
- I have some lingering questions regarding the test for the top-level code path -- please see below
- I bumped version a minor increment to 1.1.0. Happy to do otherwise if you have a strong preference; the changes to the `bin` script's interface felt like more than a patch, but I'm not sure it counts as a breaking change per se. The change to tarball behavior is substantially different but since 1.0.9 had a bug that rendered tarball handling unworkable, I don't think this should count as a breaking change either.

Here's the most controversial part: I changed the existing behavior of `test/bin.js`. I'd like to talk about whether additional revisions are appropriate. Specifically: the previous test simply `require()`d the bin script. Failure could occur in the event of a syntax error. The act of requiring kicked off all the default network requests associated with invoking the script, but mocha's default test timeout (2000ms) killed most of these off before they completed, and their result wasn't checked anyway.

In order to unit test the tarball functionality I had to change the exported interface to the script (otherwise my new tests would kick off the full-dataset download process in the course of hooking into specific methods). In for a penny/in for a pound, I put a basic async wrapper around the script's primary codepath, tested for completion of the download loop and adjusted mocha's timeout.

I am generally opposed writing tests that make network requests; but if we do make them, we should probably check their result. The project status quo makes those requests but aborts them and doesn't check the output. This strikes me as the worst of both worlds.

So! I'd like to suggest two options:
- add some checks for completeness of downloaded files
- stub out the `http` object further to handle the specific requests made in the first test, moving these tests fully offline

I think the second option is preferable, but am eager to hear what you think @msimerson.

### CODE REVIEW

**Comment by sbma44** [comment]
I'll throw out wishlist items that I'd like to implement but had decided not to in order not to rock the boat:
- cleaning up `console.log/error`s in the bin script, esp commented-out ones
- moving off mocha to `tape`

if either of these is appealing to you please say the word but otherwise I will not bother

**Comment by msimerson** [comment]
> I am generally opposed writing tests that make network requests;

Agreed.

> cleaning up console.log/errors in the bin script, esp commented-out ones

My preference is to drop in a little `lib/console` that the app and libraries all load and which is just a thin wrapper around `console`.  When all the logging entries are logged via that library, it's dead simple to hush all the logs in testing and also dead simple to pass in another logger library that suits local policies and needs.

> moving off mocha to tape

Thanks, but no thanks. I think we've found another area where we'll have to agree to disagree. 

More to follow...

**Comment by msimerson** [comment]
Also, please nuke package-lock.json and add it to .gitignore. I know the arguments for-against it and haven't found the benefits to exceed the costs. Especially for packages with minimal/no dependencies. ;)

**Comment by msimerson** [comment]
> stub out the http object further to handle the specific requests made in the first test, moving these tests fully offline

This is the preferable of the two options. However, other options are available. Moving the functions from bin/maxmind-geolite-mirror into a library, and then testing that library seems like a reasonable choice. Then just omit the test/bin.js entirely. After all, it'll have only a tiny bit of code like this:

```js
const config = require('../lib/config');
const mirror = require('../lib/mm-geolite-mirror');

mirror.doOne(config.geoIpDbs.shift(), function iterator () {
    if (!config.geoIpDbs.length) return;
    mirror.doOne(config.geoIpDbs.shift(), iterator);
})
```

Someday it would be better to use async/await in a forEach loop instead of my "poor man's async", but for now I know someone still wants node 4 support.

**Comment by sbma44** [comment]
OK, `tape` aside I think we're in sync. I will
- [ ] nuke package-lock.json (I also hate these things, sorry to have accidentally committed it--I didn't realize `npm version` automatically added it as part of the commit it makes)
- [ ] move the content of the bin script into a library and unit test
- [ ] implement a lightweight `lib/console.js`

I'm happy to punt on the "poor man's async" questions for now but will note that this code is actually a pretty good candidate for parallelism -- you can usually eke out some speed by running multiple http downloads at once, even when it's all going to the same host.

**Comment by msimerson** [comment]
>  this code is actually a pretty good candidate for parallelism

Not if you want the MaxMind folks to be happy with my script. I seem to recall reading somewhere (either on their site or in my email correspondence with them) that they frown on that, and have CDN policies in place to smack down abusers.

Were that not the case, that code would almost certainly be a Promises.all() already.

File: package.json [hunk 1]
```diff
@@ -38,7 +38,8 @@
     "contributors": "*",
     "eslint": "*",
     "mocha": "*",
-    "rewire": "*"
+    "rewire": "*",
+    "tmp": "^0.0.33"
```
Comment (msimerson) [inline package.json]: Also, please strip the version of this dependency. Odds are we almost always want the latest versions for DevDeps, and if it breaks the build it'll get caught by greenkeeper.

Comment (sbma44) [reply package.json]: ok -- haven't run into this approach before but stripping for consistency with the status quo

Comment (msimerson) [reply package.json]: The only people it'll inconvenience if a newer version of a dep breaks something is the developers, right? So why make them manually bump versions and test each time a dev dep is released? Especially when we've got tools like greenkeeper that can automatically do that for us....

Comment (msimerson) [reply package.json]: If you look [here](https://github.com/msimerson/maxmind-geolite-mirror/branches), you'll see the branches that greenkeeper automatically creates whenever a dependency releases a new version. If the tests fail, then it opens an issue.

Comment (sbma44) [reply package.json]: Yeah, this makes sense. Loose version requirements give me the willies in general but in this context I agree that there's no security issue and the potential for breakage in a dev environment is really just a matter of preference; happy to defer.

**Comment by sbma44** [comment]
Thanks @msimerson, I appreciate the quick replies and thoughtful feedback. Are you comfortable with the current code coverage level? It superficially looks like a regression but I think this doesn't reflect the actual change in coverage/test thoroughness. If you're ok with it, I think you'll need to be the one to change the codecov config to allow me to merge this.

**Comment by msimerson** [comment]
> Are you comfortable with the current code coverage level?

You caught me looking at the "misses" and seeing what's needed to get it back to 100%.  Although 100% is a tidy goal, I don't let perfection get in the way of better, and this PR certainly makes it better.

**Comment by sbma44** [comment]
Great. I'm unable to merge -- if you could do so and drop a comment here when you've had a chance to publish this new version to npm, I'd be grateful.

**Comment by msimerson** [comment]
````
$ npm publish
+ maxmind-geolite-mirror@1.1.0
````

**Pull request merged**